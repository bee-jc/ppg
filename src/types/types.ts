export interface Episode {
  id: number;
  name: string;
  image: { original: string; medium?: string };
  summary: string;
  season: number;
  imageUrl?: string;
}

export interface Season {
  id: number;
  url: string;
  number: number;
  name: string;
  episodeOrder: number;
  premiereDate: string | null;
  endDate: string | null;
  network: {
    id: number;
    name: string;
    country: {
      name: string;
      code: string;
      timezone: string;
    };
  };
  webChannel: any;
  image: {
    original: string | null;
  } | null;
  summary: string | null;
}

export interface Show {
  id: number;
  name: string;
  image: { original: string };
  summary: string;
}

export interface EpisodeState {
  details: Episode;
  loading: boolean;
  error: string | null;
}

export interface ShowState {
  details: Show;
  episodes: Episode[];
  seasons: Season[];
  loading: boolean;
}

export interface SeasonWithEpisodes extends Season {
  episodes: Episode[];
}
