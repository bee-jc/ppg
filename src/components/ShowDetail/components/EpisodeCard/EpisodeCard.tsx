import React from 'react';
import { Episode } from '../../../../types/types';
import {
  EpisodeImage,
  EpisodeTitle,
  EpisodeDescription,
  EpisodeImageContainer,
  ImagePlaceholder,
  EpisodeInfo,
  EpisodeLink,
} from '../../../../styles';

const EpisodeCard: React.FC<{ episode: Episode }> = ({ episode }) => (
  <EpisodeLink to={`/episode/${episode.id}`}>
    <EpisodeImageContainer>
      {episode.image?.medium ? (
        <EpisodeImage src={episode.image.medium} alt={episode.name} />
      ) : (
        <ImagePlaceholder>No image available</ImagePlaceholder>
      )}
    </EpisodeImageContainer>
    <EpisodeInfo>
      <EpisodeTitle>{episode.name}</EpisodeTitle>
      <EpisodeDescription>
        {episode.summary
          ? episode.summary.replace(/(<([^>]+)>)/gi, '')
          : 'No summary available.'}
      </EpisodeDescription>
    </EpisodeInfo>
  </EpisodeLink>
);

export default EpisodeCard;
