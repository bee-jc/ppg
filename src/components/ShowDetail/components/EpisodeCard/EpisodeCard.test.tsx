import React from 'react';
import { render, screen } from '@testing-library/react';
import EpisodeCard from './EpisodeCard';
import { Episode } from '../../../../types/types';
import { MemoryRouter } from 'react-router-dom';
import { create } from 'react-test-renderer';

describe('EpisodeCard', () => {
  const defaultEpisode: Episode = {
    id: 1,
    name: 'Test Episode',
    image: {
      original: 'http://example.com/image.jpg',
      medium: 'http://example.com/medium.jpg',
    },
    summary: 'This is a test summary.',
    season: 1,
  };

  it('should render an image when the episode has an image URL', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <EpisodeCard episode={defaultEpisode} />
      </MemoryRouter>
    );

    expect(screen.getByAltText('Test Episode')).toHaveAttribute(
      'src',
      'http://example.com/medium.jpg'
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a placeholder when the episode does not have an image URL', () => {
    const noImageEpisode: Episode = {
      ...defaultEpisode,
      image: { medium: '', original: '' },
    };

    render(
      <MemoryRouter>
        <EpisodeCard episode={noImageEpisode} />
      </MemoryRouter>
    );

    expect(screen.getByText('No image available')).toBeInTheDocument();
  });

  it('should render the episode title correctly', () => {
    render(
      <MemoryRouter>
        <EpisodeCard episode={defaultEpisode} />
      </MemoryRouter>
    );

    expect(screen.getByText('Test Episode')).toBeInTheDocument();
  });

  it('should render the episode summary correctly', () => {
    render(
      <MemoryRouter>
        <EpisodeCard episode={defaultEpisode} />
      </MemoryRouter>
    );

    expect(screen.getByText(defaultEpisode.summary)).toBeInTheDocument();
  });

  it('should render a fallback text when there is no episode summary', () => {
    const noSummaryEpisode: Episode = {
      ...defaultEpisode,
      summary: '',
    };

    render(
      <MemoryRouter>
        <EpisodeCard episode={noSummaryEpisode} />
      </MemoryRouter>
    );

    expect(screen.getByText('No summary available.')).toBeInTheDocument();
  });

  it('should have the correct link to the episode page', () => {
    render(
      <MemoryRouter>
        <EpisodeCard episode={defaultEpisode} />
      </MemoryRouter>
    );

    expect(screen.getByRole('link')).toHaveAttribute('href', '/episode/1');
  });

  // Snapshot test
  it('should match the snapshot', () => {
    const tree = create(
      <MemoryRouter>
        <EpisodeCard episode={defaultEpisode} />
      </MemoryRouter>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
