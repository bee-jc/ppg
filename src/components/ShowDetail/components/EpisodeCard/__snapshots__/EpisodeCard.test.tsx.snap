// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeCard should match the snapshot 1`] = `
<a
  className="css-3tqedt"
  href="/episode/1"
  onClick={[Function]}
>
  <div
    className="css-1v8ruy6"
  >
    <img
      alt="Test Episode"
      className="css-5y3q7n"
      src="http://example.com/medium.jpg"
    />
  </div>
  <div
    className="css-14mp7s"
  >
    <div
      className="css-tqpnt9"
    >
      Test Episode
    </div>
    <p
      className="css-juyb0m"
    >
      This is a test summary.
    </p>
  </div>
</a>
`;

exports[`EpisodeCard should render an image when the episode has an image URL 1`] = `
<DocumentFragment>
  <a
    class="css-3tqedt"
    href="/episode/1"
  >
    <div
      class="css-1v8ruy6"
    >
      <img
        alt="Test Episode"
        class="css-5y3q7n"
        src="http://example.com/medium.jpg"
      />
    </div>
    <div
      class="css-14mp7s"
    >
      <div
        class="css-tqpnt9"
      >
        Test Episode
      </div>
      <p
        class="css-juyb0m"
      >
        This is a test summary.
      </p>
    </div>
  </a>
</DocumentFragment>
`;
